File: reverb/cc/ops/local_table_dataset.cc
// Copyright 2019 DeepMind Technologies Limited.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <tensorflow/core/framework/tensor_util.h>
#include "tensorflow/core/framework/dataset.h"

#include "reverb/cc/local_table_sampler.h"
#include "reverb/cc/platform/logging.h"
#include "reverb/cc/local_table_sampler.h"
#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/tensor_shape.h"
#include "tensorflow/core/framework/types.h"
#include "reverb/cc/tensor_compression.h"

namespace deepmind {
namespace reverb {
namespace {

using ::tensorflow::errors::Cancelled;
using ::tensorflow::errors::FailedPrecondition;
using ::tensorflow::errors::InvalidArgument;
using ::tensorflow::errors::Unimplemented;

REGISTER_OP("ReverbLocalTableDataset")
    .Input("table: int64")
    .Attr("sequence_length: int = -1")
    .Attr("emit_timesteps: bool = true")
    .Attr("flexible_batch_size: int = -1")
    .Attr("rate_limiter_timeout_ms: int = -1")
    .Attr("dtypes: list(type) >= 1")
    .Attr("shapes: list(shape) >= 1")
    .Output("dataset: variant")
    .SetIsStateful()
    .SetShapeFn(tensorflow::shape_inference::ScalarShape)
    .Doc(R"doc(
Stream samples from table `table`. The pointer to the table is passed in as
 an int64.

`dtypes` and `shapes` must match the type and shape of a single "timestep"
within sampled sequences. That is, (key, priority, table_size, ...data passed to
`Writer::Append` at insertion time). This is the type and shape of
tensors returned by `GetNextTimestep`.

sequence_length: (Defaults to -1, i.e unknown) The number of timesteps in
the samples. If set then the length of the received samples are checked against
this value.

`emit_timesteps` (defaults to true) determines whether individual timesteps or
complete sequences should be returned from the iterators. When set to false
(i.e return sequences), `shapes` must have dim[0] equal to `sequence_length`.
Emitting complete samples is more efficient as it avoids the memcopies involved
in splitting up a sequence and then batching it up again.

`flexible_batch_size` (defaults to -1, i.e auto selected) is the maximum number
of items to sampled from `Table` with a single call. Values > 1 enables
`Table::SampleFlexibleBatch` to return more than one item (but no more than
`flexible_batch_size`) in a single call without releasing the table lock iff
the rate limiter allows it.

`rate_limiter_timeout_ms` (defaults to -1, i.e. never time out) is the number of
milliseconds an iterator should wait for new data from the sampler before timing
out. This can be useful, e.g., when the Reverb server receives data in
collection stages - and a dataset iterator should stop when no new data is
available for a while. If `rate_limiter_timeout_ms >= 0`, an iterator that waits
for data longer than this will close and mark the input sequence as finished.
Note that the timeout behavior depends on the Table's rate limiter. For example,
the table may contain data, but the rate limiter may pause sampling - and this
can cause a timeout to occur.
)doc");

class ReverbLocalTableDatasetOp : public tensorflow::data::DatasetOpKernel {
 public:
  explicit ReverbLocalTableDatasetOp(tensorflow::OpKernelConstruction* ctx)
      : tensorflow::data::DatasetOpKernel(ctx) {
    OP_REQUIRES_OK(ctx, ctx->GetAttr("sequence_length", &sequence_length_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("emit_timesteps", &emit_timesteps_));
    tensorflow::int64 rate_limiter_timeout_ms;
    OP_REQUIRES_OK(
        ctx, ctx->GetAttr("flexible_batch_size",
                          &sampler_options_.flexible_batch_size));
    OP_REQUIRES_OK(
        ctx, ctx->GetAttr("rate_limiter_timeout_ms", &rate_limiter_timeout_ms));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("dtypes", &dtypes_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("shapes", &shapes_));

    sampler_options_.rate_limiter_timeout =
        Int64MillisToNonnegativeDuration(rate_limiter_timeout_ms);

    if (!emit_timesteps_) {
      for (int i = 0; i < shapes_.size(); i++) {
        OP_REQUIRES(ctx, shapes_[i].dims() != 0,
                    InvalidArgument(
                        "When emit_timesteps is false, all elements of shapes "
                        "must have "
                        "dim[0] = sequence_length (",
                        sequence_length_, "). Element ", i,
                        " of flattened shapes has rank 0 and thus no dim[0]."));

        OP_REQUIRES(ctx, shapes_[i].dim_size(0) == sequence_length_,
                    InvalidArgument("When emit_timesteps is false, all "
                                    "elements of shapes must have "
                                    "dim[0] = sequence_length (",
                                    sequence_length_, "). Element ", i,
                                    " of flattened shapes has dim[0] = ",
                                    shapes_[i].dim_size(0), "."));
      }
    }
  }

  void MakeDataset(tensorflow::OpKernelContext* ctx,
                   tensorflow::data::DatasetBase** output) override {
    tensorflow::int64 table;
    OP_REQUIRES_OK(ctx,
                   tensorflow::data::ParseScalarArgument<tensorflow::int64>(
                           ctx, "table", &table));

    *output = new Dataset(ctx, reinterpret_cast<Table*>(table),
                          dtypes_, shapes_, sampler_options_,
                          sequence_length_, emit_timesteps_);
  }

 private:
  class Dataset : public tensorflow::data::DatasetBase {
   public:
    Dataset(tensorflow::OpKernelContext* ctx, Table* table,
            tensorflow::DataTypeVector dtypes,
            std::vector<tensorflow::PartialTensorShape> shapes,
            const LocalTableSampler::Options& sampler_options,
            int sequence_length, bool emit_timesteps)
        : tensorflow::data::DatasetBase(tensorflow::data::DatasetContext(ctx)),
          table_(table),
          dtypes_(std::move(dtypes)),
          shapes_(std::move(shapes)),
          sampler_options_(sampler_options),
          sequence_length_(sequence_length),
          emit_timesteps_(emit_timesteps) {}

    std::unique_ptr<tensorflow::data::IteratorBase> MakeIteratorInternal(
        const std::string& prefix) const override {
      return absl::make_unique<Iterator>(
          tensorflow::data::DatasetIterator<Dataset>::Params{
              this, absl::StrCat(prefix, "::ReverbLocalTableDataset")},
          table_, sampler_options_, sequence_length_,
          emit_timesteps_, dtypes_, shapes_);
    }

    const tensorflow::DataTypeVector& output_dtypes() const override {
      return dtypes_;
    }

    const std::vector<tensorflow::PartialTensorShape>& output_shapes()
        const override {
      return shapes_;
    }

    std::string DebugString() const override {
      return "ReverbDatasetOp::Dataset";
    }

    tensorflow::Status CheckExternalState() const override {
      return FailedPrecondition(DebugString(), " depends on external state.");
    }

   protected:
    tensorflow::Status AsGraphDefInternal(
        tensorflow::data::SerializationContext* ctx, DatasetGraphDefBuilder* b,
        tensorflow::Node** output) const override {
      tensorflow::AttrValue sequence_length_attr;
      tensorflow::AttrValue emit_timesteps_attr;
      tensorflow::AttrValue flexible_batch_size_attr;
      tensorflow::AttrValue rate_limiter_timeout_ms_attr;
      tensorflow::AttrValue dtypes_attr;
      tensorflow::AttrValue shapes_attr;

      tensorflow::Node* table = nullptr;

      TF_RETURN_IF_ERROR(
              b->AddScalar<tensorflow::int64>((int64_t)table_, &table));

      b->BuildAttrValue(sequence_length_, &sequence_length_attr);
      b->BuildAttrValue(emit_timesteps_, &emit_timesteps_attr);
      b->BuildAttrValue(sampler_options_.flexible_batch_size,
          &flexible_batch_size_attr);
      b->BuildAttrValue(
          static_cast<tensorflow::int64>(NonnegativeDurationToInt64Millis(
              sampler_options_.rate_limiter_timeout)),
          &rate_limiter_timeout_ms_attr);
      b->BuildAttrValue(dtypes_, &dtypes_attr);
      b->BuildAttrValue(shapes_, &shapes_attr);

      TF_RETURN_IF_ERROR(b->AddDataset(
          this,
          /*inputs=*/{table},
          /*attrs=*/
          {
              {"sequence_length", sequence_length_attr},
              {"emit_timesteps", emit_timesteps_attr},
              {"flexible_batch_size", flexible_batch_size_attr},
              {"rate_limiter_timeout_ms", rate_limiter_timeout_ms_attr},
              {"dtypes", dtypes_attr},
              {"shapes", shapes_attr},
          },
          output));

      return tensorflow::Status::OK();
    }

   private:
    class Iterator : public tensorflow::data::DatasetIterator<Dataset> {
     public:
      explicit Iterator(
          const Params& params, Table* table,
          const LocalTableSampler::Options& sampler_options,
          int sequence_length, bool emit_timesteps,
          const tensorflow::DataTypeVector& dtypes,
          const std::vector<tensorflow::PartialTensorShape>& shapes)
          : DatasetIterator<Dataset>(params),
            table_(table),
            sampler_options_(sampler_options),
            sequence_length_(sequence_length),
            emit_timesteps_(emit_timesteps),
            dtypes_(dtypes),
            shapes_(shapes),
            step_within_sample_(0) { }

        tensorflow::Status Initialize(
          tensorflow::data::IteratorContext* ctx) override {
        // If sequences are emitted then the all shapes will start with the
        // sequence length. The validation expects the shapes of a single
        // timestep so if sequences are emitted then we need to trim the leading
        // dim on all shapes before validating it.
        auto validation_shapes = shapes_;
        if (!emit_timesteps_) {
          for (auto& shape : validation_shapes) {
            shape.RemoveDim(0);
          }
        }

        TF_RETURN_IF_ERROR(sampler_options_.Validate());
        // TODO(piyushk): Note that dtypes and shapes validation still needs
        // to be added.
        sampler_ = absl::make_unique<LocalTableSampler>(
            table_, sampler_options_);

        return tensorflow::Status::OK();
      }

      tensorflow::Status GetNextInternal(
          tensorflow::data::IteratorContext* ctx,
          std::vector<tensorflow::Tensor>* out_tensors,
          bool* end_of_sequence) override {
        REVERB_CHECK(sampler_.get() != nullptr) << "Initialize was not called?";

        tensorflow::Status status;
        if (emit_timesteps_) {
          bool last_timestep = false;
          status = sampler_->GetNextTimestep(out_tensors, &last_timestep);

          step_within_sample_++;

          if (last_timestep && sequence_length_ > 0 &&
              step_within_sample_ != sequence_length_) {
            return InvalidArgument(
                "Received sequence of invalid length. Expected ",
                sequence_length_, " steps, got ", step_within_sample_);
          }
          if (step_within_sample_ == sequence_length_ && !last_timestep) {
            return InvalidArgument(
                "Receieved sequence did not terminate after expected number of "
                "steps (",
                sequence_length_, ").");
          }
          if (last_timestep) {
            step_within_sample_ = 0;
          }
        } else {
          status = sampler_->GetNextSample(out_tensors);
        }

        if (status.ok()) {
          *end_of_sequence = false;
          return status;
        } else if (tensorflow::errors::IsDeadlineExceeded(status) &&
                   sampler_options_.rate_limiter_timeout <
                       absl::InfiniteDuration() &&
                   status.error_message().find("Rate Limiter") !=
                       std::string::npos) {
          // TODO(157580783): Move the error string above to a common library.
          *end_of_sequence = true;
          return tensorflow::Status::OK();
        } else {
          return status;
        }
      }

     protected:
      tensorflow::Status SaveInternal(
          tensorflow::data::SerializationContext* ctx,
          tensorflow::data::IteratorStateWriter* writer) override {
        return Unimplemented("SaveInternal is currently not supported");
      }

      tensorflow::Status RestoreInternal(
          tensorflow::data::IteratorContext* ctx,
          tensorflow::data::IteratorStateReader* reader) override {
        return Unimplemented("RestoreInternal is currently not supported");
      }

     private:
      Table* table_;
      const LocalTableSampler::Options sampler_options_;
      const int sequence_length_;
      const bool emit_timesteps_;
      const tensorflow::DataTypeVector& dtypes_;
      const std::vector<tensorflow::PartialTensorShape>& shapes_;
      std::unique_ptr<LocalTableSampler> sampler_;
      int step_within_sample_;
    };  // Iterator.

    Table* table_;
    const tensorflow::DataTypeVector dtypes_;
    const std::vector<tensorflow::PartialTensorShape> shapes_;
    const LocalTableSampler::Options sampler_options_;
    const int sequence_length_;
    const bool emit_timesteps_;
  };  // Dataset.

  LocalTableSampler::Options sampler_options_;
  int sequence_length_;
  bool emit_timesteps_;
  tensorflow::DataTypeVector dtypes_;
  std::vector<tensorflow::PartialTensorShape> shapes_;

  TF_DISALLOW_COPY_AND_ASSIGN(ReverbLocalTableDatasetOp);
};

REGISTER_KERNEL_BUILDER(
    Name("ReverbLocalTableDataset").Device(tensorflow::DEVICE_CPU),
    ReverbLocalTableDatasetOp
);

}  // namespace
}  // namespace reverb
}  // namespace deepmind

File: reverb/local_table_dataset.py
# Lint as: python3
# Copyright 2019 DeepMind Technologies Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""TFClient provides tf-ops for interacting with Reverb."""

from typing import Any, List, Optional, Union

from reverb import client as reverb_client, server as reverb_server
from reverb import replay_sample
import tensorflow.compat.v1 as tf
import tree

from reverb.cc.ops import gen_local_table_dataset_op


class ReplayLocalTableDataset(tf.data.Dataset):
  """A tf.data.Dataset which samples timesteps from the ReverbService.

  Note: The dataset returns `ReplaySample` where `data` with the structure of
  `dtypes` and `shapes`.

  Note: Uses of Python lists are converted into tuples as nest used by the
  tf.data API doesn't have good support for lists.

  Timesteps are streamed through the dataset as follows:

    1. Does an active prioritized item exists?
         - Yes: Go to 3
         - No: Go to 2.
    2. Sample a prioritized item from `table` using its sample-function and set
       the item as "active". Go to 3.
    3. Yield the next timestep within the active prioritized item. If the
       timestep was the last one within the item, clear its "active" status.

  This allows for items of arbitrary length to be streamed with limited memory.
  """

  def __init__(self,
               table: Union[reverb_server.Table, tf.Tensor],
               dtypes: Any,
               shapes: Any,
               sequence_length: Optional[int] = None,
               emit_timesteps: bool = True,
               flexible_batch_size: int = -1,
               rate_limiter_timeout_ms: int = -1):
    """Constructs a new ReplayLocalTableDataset.

    Args:
      table: Probability table to sample from.
      dtypes: Dtypes of the data output. Can be nested.
      shapes: Shapes of the data output. Can be nested.
      sequence_length: (Defaults to None, i.e unknown) The number of timesteps
        that each sample consists of. If set then the length of samples received
        from the server will be validated against this number.
      emit_timesteps: (Defaults to True) If set, timesteps instead of full
        sequences are returned from the dataset. Returning sequences instead of
        timesteps can be more efficient as the memcopies caused by the splitting
        and batching of tensor can be avoided. Note that if set to False then
        then all `shapes` must have dim[0] equal to `sequence_length`.
      flexible_batch_size: (Defaults to -1, i.e auto selected) is the maximum number
        of items to sampled from `Table` with a single call. Values > 1 enables
        `Table::SampleFlexibleBatch` to return more than one item (but no more than
        `flexible_batch_size`) in a single call without releasing the table lock iff
        the rate limiter allows it.
      rate_limiter_timeout_ms: (Defaults to -1: infinite).  Timeout
        (in milliseconds) to wait on the rate limiter when sampling from the
        table. If `rate_limiter_timeout_ms >= 0`, this is the timeout passed to
        `Table::Sample` describing how long to wait for the rate limiter to
        allow sampling. The first time that a request times out (across any of
        the workers), the Dataset iterator is closed and the sequence is
        considered finished.

    Raises:
      ValueError: If `dtypes` and `shapes` don't share the same structure.
      ValueError: If `sequence_length` is not a positive integer or None.
      ValueError: If `emit_timesteps is False` and not all items in `shapes` has
        `sequence_length` as its leading dimension.
      ValueError: If `flexible_batch_size < -1`.
      ValueError: If `rate_limiter_timeout_ms < -1`.
    """
    tree.assert_same_structure(dtypes, shapes, False)
    if sequence_length is not None and sequence_length < 1:
      raise ValueError(
          'sequence_length (%s) must be None or a positive integer' %
          sequence_length)
    if flexible_batch_size < -1:
        raise ValueError('flexible_batch_size (%d) must be an integer >= -1' %
                         flexible_batch_size)
    if rate_limiter_timeout_ms < -1:
      raise ValueError('rate_limiter_timeout_ms (%d) must be an integer >= -1' %
                       rate_limiter_timeout_ms)

    # Add the info fields.
    dtypes = replay_sample.ReplaySample(replay_sample.SampleInfo.tf_dtypes(),
                                        dtypes)
    shapes = replay_sample.ReplaySample(
        replay_sample.SampleInfo(
            tf.TensorShape([sequence_length] if not emit_timesteps else []),
            tf.TensorShape([sequence_length] if not emit_timesteps else []),
            tf.TensorShape([sequence_length] if not emit_timesteps else []),
            tf.TensorShape([sequence_length] if not emit_timesteps else [])),
        shapes)

    # If sequences are to be emitted then all shapes must specify use
    # sequence_length as their batch dimension.
    if not emit_timesteps:

      def _validate_batch_dim(path: str, shape: tf.TensorShape):
        if (not shape.ndims
            or tf.compat.dimension_value(shape[0]) != sequence_length):
          raise ValueError(
              'All items in shapes must use sequence_range (%s) as the leading '
              'dimension, but "%s" has shape %s' %
              (sequence_length, path[0], shape))

      tree.map_structure_with_path(_validate_batch_dim, shapes.data)

    # The tf.data API doesn't fully support lists so we convert all uses of
    # lists into tuples.
    dtypes = _convert_lists_to_tuples(dtypes)
    shapes = _convert_lists_to_tuples(shapes)

    self._table = table
    self._dtypes = dtypes
    self._shapes = shapes
    self._sequence_length = sequence_length
    self._emit_timesteps = emit_timesteps
    self._flexible_batch_size = flexible_batch_size
    self._rate_limiter_timeout_ms = rate_limiter_timeout_ms

    if _is_tf1_runtime():
      # Disabling to avoid errors given the different tf.data.Dataset init args
      # between v1 and v2 APIs.
      # pytype: disable=wrong-arg-count
      super().__init__()
    else:
      # DatasetV2 requires the dataset as a variant tensor during init.
      super().__init__(self._as_variant_tensor())
      # pytype: enable=wrong-arg-count

  # TODO(piyushk): Currently unsupported.
  # @classmethod
  # def from_table_signature(cls,
  #                          table: reverb_server.Table,
  #                          sequence_length: Optional[int] = None,
  #                          emit_timesteps: bool = True,
  #                          flexible_batch_size: int = -1,
  #                          rate_limiter_timeout_ms: int = -1,
  #                          get_signature_timeout_secs: Optional[int] = None,):
  #   """Constructs a ReplayLocalTableDataset using the table's signature to infer specs.
  #
  #   Note: The signature must be provided to `Table` at construction. See
  #   `Table.__init__` (./server.py) for more details.
  #
  #   Args:
  #     table: Table to read the signature and sample from.
  #     sequence_length: See __init__ for details.
  #     emit_timesteps: See __init__ for details.
  #     flexible_batch_size: See __init__ for details.
  #     rate_limiter_timeout_ms: See __init__ for details.
  #     get_signature_timeout_secs: Timeout in seconds to wait for server to
  #       respond when fetching the table signature. By default no timeout is set
  #       and the call will block indefinetely if the server does not respond.
  #
  #   Returns:
  #     ReplayLocalTableDataset using the specs defined by the table signature to build
  #       `shapes` and `dtypes`.
  #
  #   Raises:
  #     ValueError: If `table` does not exist on server at `server_address`.
  #     ValueError: If `table` does not have a signature.
  #     errors.DeadlineExceededError: If `get_signature_timeout_secs` provided and
  #       exceeded.
  #     ValueError: See __init__.
  #   """
  #   info = client.server_info(get_signature_timeout_secs)
  #   if table not in info:
  #     raise ValueError(
  #         f'Server at {server_address} does not contain any table named '
  #         f'{table}. Found: {", ".join(sorted(info.keys()))}.')
  #
  #   if not info[table].signature:
  #     raise ValueError(
  #         f'Table {table} at {server_address} does not have a signature.')
  #
  #   shapes = tree.map_structure(lambda x: x.shape, info[table].signature)
  #   dtypes = tree.map_structure(lambda x: x.dtype, info[table].signature)
  #
  #   if not emit_timesteps:
  #     batch_dim = tf.TensorShape([sequence_length])
  #     shapes = tree.map_structure(batch_dim.concatenate, shapes)
  #
  #   return cls(
  #       table=table,
  #       shapes=shapes,
  #       dtypes=dtypes,
  #       num_workers_per_iterator=num_workers_per_iterator,
  #       sequence_length=sequence_length,
  #       emit_timesteps=emit_timesteps,
  #       rate_limiter_timeout_ms=rate_limiter_timeout_ms)

  def _as_variant_tensor(self):
    return gen_local_table_dataset_op.reverb_local_table_dataset(
        table=self._table.get_ptr(),
        dtypes=tree.flatten(self._dtypes),
        shapes=tree.flatten(self._shapes),
        emit_timesteps=self._emit_timesteps,
        sequence_length=self._sequence_length or -1,
        flexible_batch_size=self._flexible_batch_size,
        rate_limiter_timeout_ms=self._rate_limiter_timeout_ms)

  def _inputs(self) -> List[Any]:
    return []

  @property
  def element_spec(self) -> Any:
    return tree.map_structure(tf.TensorSpec, self._shapes, self._dtypes)


def _convert_lists_to_tuples(structure: Any) -> Any:
  list_to_tuple_fn = lambda s: tuple(s) if isinstance(s, list) else s
  # Traverse depth-first, bottom-up
  return tree.traverse(list_to_tuple_fn, structure, top_down=False)


def _is_tf1_runtime() -> bool:
  """Returns True if the runtime is executing with TF1.0 APIs."""
  # TODO(b/145023272): Update when/if there is a better way.
  return hasattr(tf, 'to_float')
